grype/presenter/explain/explain.go

```diff
@@ -0,0 +1,467 @@
+	MatchedOnNamespace  string
+	IndirectExplanation string
+	DirectExplanation   string
+	CPEExplanation      string
+	Locations           []explainedEvidence
+	displayPriority     int // shows how early it should be displayed; direct matches first
+}
+
+type explainedEvidence struct {
+	Location     string
+	ArtifactID   string
+	ViaVulnID    string
+	ViaNamespace string
+}
+
+type vulnerabilityExplainer struct {
+	w   io.Writer
+	doc *models.Document
+}
+
+func NewVulnerabilityExplainer(w io.Writer, doc *models.Document) VulnerabilityExplainer {
+	return &vulnerabilityExplainer{
+		w:   w,
+		doc: doc,
+	}
+}
+
+var funcs = template.FuncMap{
+	"trim": strings.TrimSpace,
+}
+
+func (e *vulnerabilityExplainer) ExplainByID(ids []string) error {
+	findings, err := Doc(e.doc, ids)
+	if err != nil {
+		return err
+	}
+	t := template.Must(template.New("explanation").Funcs(funcs).Parse(explainTemplate))
+	for _, id := range ids {
+		finding, ok := findings[id]
+		if !ok {
+			continue
+		}
+		if err := t.Execute(e.w, finding); err != nil {
+			return fmt.Errorf("unable to execute template: %w", err)
+		}
+	}
+	return nil
+}
+
+func (e *vulnerabilityExplainer) ExplainBySeverity(_ string) error {
+	return fmt.Errorf("not implemented")
+}
+
+func (e *vulnerabilityExplainer) ExplainAll() error {
+	findings, err := Doc(e.doc, nil)
+	if err != nil {
+		return err
+	}
+	t := template.Must(template.New("explanation").Funcs(funcs).Parse(explainTemplate))
+
+	return t.Execute(e.w, findings)
+}
+
+func Doc(doc *models.Document, requestedIDs []string) (Findings, error) {
+	result := make(Findings)
+	builders := make(map[string]*viewModelBuilder)
+	for _, m := range doc.Matches {
+		key := m.Vulnerability.ID
+		existing, ok := builders[key]
+		if !ok {
+			existing = newBuilder(requestedIDs)
+			builders[m.Vulnerability.ID] = existing
+		}
+		existing.WithMatch(m, requestedIDs)
+	}
+	for _, m := range doc.Matches {
+		for _, related := range m.RelatedVulnerabilities {
+			key := related.ID
+			existing, ok := builders[key]
+			if !ok {
+				existing = newBuilder(requestedIDs)
+				builders[key] = existing
+			}
+			// TODO: need to pass in info about the related vulnerability
+			existing.WithMatch(m, requestedIDs)
+		}
+	}
+	for k, v := range builders {
+		result[k] = v.Build()
+	}
+	return result, nil
+}
+
+func newBuilder(requestedIDs []string) *viewModelBuilder {
+	return &viewModelBuilder{
+		requestedIDs: requestedIDs,
+	}
+}
+
+// WithMatch adds a match to the builder
+// accepting enough information to determine whether the match is a primary match or a related match
+func (b *viewModelBuilder) WithMatch(m models.Match, userRequestedIDs []string) {
+	if b.isPrimaryAdd(m, userRequestedIDs) {
+		// Demote the current primary match to related match
+		// if it exists
+		if b.PrimaryMatch.Vulnerability.ID != "" {
+			b.WithRelatedMatch(b.PrimaryMatch)
+		}
+		b.WithPrimaryMatch(m)
+	} else {
+		b.WithRelatedMatch(m)
+	}
+}
+
+// TODO: is this still needed?
```

Comment:
Based on reading this function I think it's still needed, but maybe you have found a way to simplify the `WithMatch` branches where we don't need to care about `Primary` vs `Related` anymore?